### PR TITLE
cache: create directories/files with stricter permissions, check existing permissions.

### PR DESCRIPTION
**Pull request by klihub**
Spell out more explicitly the expected/preferred vs. rejected permission bits for the various cache/container data files and directories.Check that files and directories have root/root as owner/group. Violation of ownership or permission check of existing files/directories results in an error/rejection (and the failure of the related operation). Non-existent files/directories are created with the expected/preferred permissions (without an extra explicit check to see if we run as root/root).

### CODE REVIEW

File: pkg/cri/resource-manager/cache/cache.go [hunk 1]
```diff
@@ -490,6 +490,12 @@ type Cache interface {
 const (
 	// CacheVersion is the running version of the cache.
 	CacheVersion = "1"
+	// cacheFilePermissions is the unix permissions for the cache file
+	cacheFilePermissions = os.FileMode(0640)
+	// cacheDirPermissions is the unix permissions for the directory of the cache file
+	cacheDirPermissions = os.FileMode(0710)
```
Comment (ipuustin) [inline pkg/cri/resource-manager/cache/cache.go:496]: Here the original permission was stricter... but maybe the group access is a good thing if we move later running the service as a user.

File: pkg/cri/resource-manager/cache/cache.go [hunk 2]
```diff
@@ -1110,6 +1119,37 @@ func (cch *cache) setEntry(key string, ptr, obj interface{}) error {
 	return nil
 }
 
+// checkFilePermissions check and warns if an existing file has unexpected permissions.
+func (cch *cache) checkFilePermissions(path string, perm os.FileMode, what string) {
+	if info, err := os.Stat(path); err == nil && !info.IsDir() {
+		expected := perm & os.ModePerm // deliberately ignore sticky bit in comparison
+		existing := info.Mode().Perm() & os.ModePerm
+		if existing != expected && (existing|expected) != expected {
+			cch.Warn("existing %s file %q has less strict permissions %v than expected %v",
```
Comment (ipuustin) [inline pkg/cri/resource-manager/cache/cache.go:1128]: Why not error out?

File: pkg/cri/resource-manager/cache/cache.go (hunk 1 above)
Comment (klihub) [reply pkg/cri/resource-manager/cache/cache.go:496]: Yes, that's deliberate. Maybe the commit message is a bit misleading.

File: pkg/cri/resource-manager/cache/cache.go (hunk 2 above)
Comment (klihub) [reply pkg/cri/resource-manager/cache/cache.go:1128]: With a pre-created directory, the more liberal permissions might be a deliberate choice and we should be very careful when/if we try to outsmart the sysadmin. This is what we agreed in the last Fri session, but we should probably still error out in some extreme cases, for instance if a file or directory is unexpectedly world-readable. I see if I can update this to do that in some acceptable form...

Comment (klihub) [reply pkg/cri/resource-manager/cache/cache.go:1141]: Added checking of permissions bits to reject.

File: pkg/cri/resource-manager/cache/cache.go [hunk 3]
```diff
@@ -1110,6 +1128,47 @@ func (cch *cache) setEntry(key string, ptr, obj interface{}) error {
 	return nil
 }
 
+// checkFilePermissions checks permission bits of an existing file.
+func (cch *cache) checkFilePermissions(path string, perm, rejected os.FileMode, what string) error {
+	if info, err := os.Stat(path); err == nil && !info.IsDir() {
+		expected := perm & os.ModePerm // deliberately ignore sticky bit in comparison
+		existing := info.Mode().Perm() & os.ModePerm
+		if rejected != 0 && (existing&rejected) != 0 {
```
Comment (marquiz) [inline pkg/cri/resource-manager/cache/cache.go:1136]: nit: the first check is not needed, right(?) If rejected==0 then existing&rejected==0

File: pkg/cri/resource-manager/cache/cache.go [hunk 4]
```diff
@@ -1110,6 +1128,47 @@ func (cch *cache) setEntry(key string, ptr, obj interface{}) error {
...
+			return cacheError("existing %s file %q has disallowed permissions set: %v",
+				what, path, existing&rejected)
+		}
+		if existing != expected && (existing|expected) != expected {
```
Comment (marquiz) [inline pkg/cri/resource-manager/cache/cache.go:1140]: nit: I think this logic is also somehow overcomplicated. Wouldn't `(existing|expected) != expected` be enough, because if that is true then certainly `existing != expected`?

File: pkg/cri/resource-manager/cache/cache.go [hunk 5]
```diff
@@ -1110,6 +1128,47 @@ func (cch *cache) setEntry(key string, ptr, obj interface{}) error {
...
+		expected := perm & os.ModePerm // deliberately ignore sticky bit in comparison
+		existing := info.Mode().Perm() & os.ModePerm
+		if rejected != 0 && (existing&rejected) != 0 {
+			return cacheError("existing %s directory %q has disallowed permissions set: %v",
+				what, path, existing&rejected)
+		}
+		if existing != expected && (existing|expected) != expected {
+			cch.Warn("existing %s directory %q has less strict permissions %v than expected %v",
+				what, path, existing, expected)
+		}
```
Comment (marquiz) [inline pkg/cri/resource-manager/cache/cache.go:1161]: Is this duplicated from `checkFilePermissions()` if so, could it be split out in a separate helper function?

File: pkg/cri/resource-manager/cache/cache.go (hunk 3 above)
Comment (klihub) [reply pkg/cri/resource-manager/cache/cache.go]: Split out permissions check, added missing ownership tests, and removed the extra zero-test.

File: pkg/cri/resource-manager/cache/cache.go (hunk 4 above)
Comment (klihub) [reply pkg/cri/resource-manager/cache/cache.go]: Removed extra-articulated inequality check.

File: pkg/cri/resource-manager/cache/cache.go (hunk 5 above)
Comment (klihub) [reply pkg/cri/resource-manager/cache/cache.go]: Split out/combined permission checks.

File: pkg/cri/resource-manager/cache/cache.go [hunk 6]
```diff
@@ -1110,6 +1131,84 @@ func (cch *cache) setEntry(key string, ptr, obj interface{}) error {
...
+	}
+
+	// check expected root/root ownership
+	stat, ok := info.Sys().(*syscall.Stat_t)
+	if !ok || stat == nil {
+		return true, cacheError("failed to get ownership of %s %q: can't get syscall.Stat_t",
+			what, path)
+	}
+	if stat.Uid != 0 || stat.Gid != 0 {
```
Comment (marquiz) [inline pkg/cri/resource-manager/cache/cache.go:1165]: Do we want to strictly enforce this? I think it is possible to run the container runtime as rootless, for example. Or maybe run cri-resmgr as non-root when doing introspection only, for example. WDYT?

Comment (klihub) [reply pkg/cri/resource-manager/cache/cache.go]: Hmm... I considered checking against geteuid()/getegid() then went with this. Perhaps both because I was a bit too much in SDL mode. Let's drop the user/group ownership checks altogether.
